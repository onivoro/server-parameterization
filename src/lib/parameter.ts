export abstract class Parameter<T> {
    key() {
        return this.transform(this.id());
    }

    abstract value(): T;
    protected abstract id(): string;
    protected abstract transform(id: string): string;
}