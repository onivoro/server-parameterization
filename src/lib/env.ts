import * as kebabcase from 'lodash.kebabcase';
import { Parameter } from './parameter';
const DASHES = /-/g;

export abstract class Env<T> extends Parameter<T> {
    protected transform(id: string): string {
        return kebabcase(id.replace(DASHES, '_')).toUpperCase();
    }

    value() {
        return process.env[this.key()] as any;
    }
}
