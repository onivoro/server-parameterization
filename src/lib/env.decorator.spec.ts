import "reflect-metadata";
import { Parameter } from './parameter.class';
import { Env } from './env.decorator';

let processEnvMock = {
    'SOME_STRING': 'asdf',
    'SOME_NUMBER': '337',
    'SOME_BOOL': 'false',
    'OTHER_BOOL': 'true',
    'SOME_DATE': '2020-12-20T11:31:41.378Z',
};

@Env(String, processEnvMock) export class SomeString extends Parameter<string> { }
@Env(Number, processEnvMock) export class SomeNumber extends Parameter<number> { }
@Env(Boolean, processEnvMock) export class SomeBool extends Parameter<boolean> { }
@Env(Boolean, processEnvMock) export class OtherBool extends Parameter<boolean> { }
@Env(Date, processEnvMock) export class SomeDate extends Parameter<Date> { }


describe('envDecorator', () => {
    it.each([
        new SomeString(),
        new SomeNumber(),
        new SomeBool(),
        new OtherBool(),
        new SomeDate(),
    ])('reads the conventionally-implied env var and stores both raw and typed representations thereof', (envVar) => {
        expect(envVar).toMatchSnapshot();
    });
});
