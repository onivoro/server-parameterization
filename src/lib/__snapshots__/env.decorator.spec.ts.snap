// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`envDecorator reads the conventionally-implied env var and stores both raw and typed representations thereof 1`] = `
SomeString {
  "key": "SOME_STRING",
  "source": "process.env",
  "typedValue": "asdf",
  "value": "asdf",
}
`;

exports[`envDecorator reads the conventionally-implied env var and stores both raw and typed representations thereof 2`] = `
SomeNumber {
  "key": "SOME_NUMBER",
  "source": "process.env",
  "typedValue": 337,
  "value": "337",
}
`;

exports[`envDecorator reads the conventionally-implied env var and stores both raw and typed representations thereof 3`] = `
SomeBool {
  "key": "SOME_BOOL",
  "source": "process.env",
  "typedValue": false,
  "value": "false",
}
`;

exports[`envDecorator reads the conventionally-implied env var and stores both raw and typed representations thereof 4`] = `
OtherBool {
  "key": "OTHER_BOOL",
  "source": "process.env",
  "typedValue": true,
  "value": "true",
}
`;

exports[`envDecorator reads the conventionally-implied env var and stores both raw and typed representations thereof 5`] = `
SomeDate {
  "key": "SOME_DATE",
  "source": "process.env",
  "typedValue": 2020-12-20T11:31:41.378Z,
  "value": "2020-12-20T11:31:41.378Z",
}
`;
